### PR TITLE
Merge missing Camp features into PartyMap, then 💥 the old Camp components

### PR DESCRIPTION
**Pull request by 0xdevalias**
Ensure the `PartyMap` components have all of the features from the old `Camp` components, then 💥 the old Camp components

**Note:** The bulk of this work was completed/tracked in https://github.com/sparkletown/sparkle/pull/840, along with interim reviews and related. This new PR is to reduce the cognitive effort of performing final reviews/cleanup before merging back into `staging`.

## Preview Deploy

- see URL in https://github.com/sparkletown/sparkle/pull/1080#issuecomment-762057287

## TODO

- [ ] remaining comments/issues in https://github.com/sparkletown/sparkle/pull/1080#issuecomment-762615593

### CODE REVIEW

File: src/api/bannerAdmin.ts
Comment (denisdimitrov) [inline src/api/bannerAdmin.ts:12]: Is this intentionally without async/await?

File: src/components/organisms/BannerAdmin.tsx
Comment (denisdimitrov) [inline src/components/organisms/BannerAdmin.tsx:35]: return false?

File: src/components/organisms/BannerAdmin.tsx
Comment (denisdimitrov) [inline src/components/organisms/BannerAdmin.tsx:14]: Is there a particular reason to use `ref` instead of `useState` ?

File: src/api/profile.ts
Comment (denisdimitrov) [inline src/api/profile.ts:41]: +1

File: src/api/bannerAdmin.ts
Comment (0xdevalias) [reply src/api/bannerAdmin.ts:12]: I don't think so. I would assume I didn't necessarily think about it because it's not returning a value. But it would still be good to expose the fact that it's technically a `Promise`

File: src/components/organisms/BannerAdmin.tsx
Comment (0xdevalias) [reply src/components/organisms/BannerAdmin.tsx:35]: My guess is that it's intent was to prevent the default action from the `<button>` `onClick` handler, but that doesn't work in React:

- https://reactjs.org/docs/handling-events.html
  - > Another difference is that you cannot `return false` to prevent default behavior in React. You must call `preventDefault` explicitly.

File: src/components/organisms/BannerAdmin.tsx
Comment (0xdevalias) [reply src/components/organisms/BannerAdmin.tsx:14]: We're not storing any state here, but we are getting a reference to the `<input>` element/DOM node itself:

- https://reactjs.org/docs/hooks-reference.html#useref
  - > You might be familiar with refs primarily as a way to access the DOM. If you pass a `ref` object to React with `<div ref={myRef} />`, React will set its `.current` property to the corresponding DOM node whenever that node changes.

Comment (0xdevalias) [reply src/components/organisms/BannerAdmin.tsx:14]: We're not storing any state here, but we are getting a reference to the `<input>` element/DOM node itself:

- https://reactjs.org/docs/hooks-reference.html#useref
  - > You might be familiar with refs primarily as a way to access the DOM. If you pass a `ref` object to React with `<div ref={myRef} />`, React will set its `.current` property to the corresponding DOM node whenever that node changes.
  
This is being used in the `useCallback`'d `saveBanner` function to reference `inputFieldRef.current.value`, so that the `useCallback` memo doesn't change on every update to the input field (because a ref always has a stable reference for comparisons, only it's `.current` changes reference)

**Comment by 0xdevalias** [comment]
- [ ] Resolved?

Other comments from @denisdimitrov:
> Overall, looks good. I left some comments and clicked around on the preview link. Some of the styles are not correct and there are some transitions missing. For example, when hovering a room it changes the size directly without the transition animation. The blur is missing on the background of the name/count. And the border radius for the room image is missing.

File: src/api/profile.ts
Comment (0xdevalias) [reply src/api/profile.ts]: fixed in https://github.com/sparkletown/sparkle/pull/1080/commits/5379fe84c2ba33c352178141ac8473e8108c2788

File: src/components/organisms/BannerAdmin.tsx
Comment (0xdevalias) [reply src/components/organisms/BannerAdmin.tsx]: fixed in https://github.com/sparkletown/sparkle/pull/1080/commits/d712013d6dc59ef2a8b177f0050bcf5eb30510ce

File: src/api/bannerAdmin.ts
Comment (0xdevalias) [reply src/api/bannerAdmin.ts]: fixed in https://github.com/sparkletown/sparkle/pull/1080/commits/7d64d0f166ee714ac276a4c885ed2c8b1af69c15

File: src/components/templates/PartyMap/components/Map/Map.scss
Comment (0xdevalias) [inline src/components/templates/PartyMap/components/Map/Map.scss:200]: Note: The relevant room/etc styles from here were moved into `src/components/templates/PartyMap/components/Map/MapRoom.scss`

File: src/hooks/useKeyboardControls.ts
Comment (0xdevalias) [inline src/hooks/useKeyboardControls.ts:27]: [OOS]
> - use `e.preventDefault()` or `return false` or similar in the keyboard handlers (check mousetrap docs) so we don't scroll on arrow key presses
> - we may also need some kind of 'centre me on the map' logic somewhere when we do this too? For large zoomed maps.
> - implement `bindRef` using `useRef()` or similar (in PartyMap?) then remove all `withGlobalBind` (default is `false`)

File: src/pages/VideoAdmin/VideoAdmin.tsx
Comment (0xdevalias) [inline src/pages/VideoAdmin/VideoAdmin.tsx:8]: [OOS]
> `@debt` refactor BannerAdmin and this to DRY the common parts.
>   The only difference in pattern is the update function being called + placeholder/etc labels

File: src/types/venues.ts
Comment (0xdevalias) [inline src/types/venues.ts:13]: > should JazzBarVenue be added to this?

File: src/types/venues.ts
Comment (0xdevalias) [inline src/types/venues.ts:102]: > `@debt` which of these params are exactly the same as on Venue? Can we simplify this?

File: src/types/venues.ts
Comment (0xdevalias) [inline src/types/venues.ts:161]: [OOS]
> `@debt` landingPageConfig should probably be undefined, or is it guaranteed to exist everywhere?

```suggestion
  landingPageConfig?: VenueLandingPageConfig;
```

File: src/types/venues.ts
Comment (0xdevalias) [inline src/types/venues.ts:233]: > refactor this somewhere more useful, maybe utils?

File: src/utils/useLocationUpdateEffect.ts
Comment (0xdevalias) [inline src/utils/useLocationUpdateEffect.ts:50]: [OOS]
> rename this trackLocationEntered?

File: src/components/templates/PartyMap/components/Map/hooks/usePartygoersOverlay.tsx
Comment (0xdevalias) [inline src/components/templates/PartyMap/components/Map/hooks/usePartygoersOverlay.tsx:47]: > - `@debt` partygoers can be undefined because our types are broken so check explicitly
> - `@debt` workaround, sometimes partygoers are duplicated but the new ones don't have id's

File: src/components/templates/PartyMap/components/Map/Map.tsx
Comment (0xdevalias) [inline src/components/templates/PartyMap/components/Map/Map.tsx:19]: > `@debt` refactor these hooks into somewhere more sensible

This comment was made when we were going to be sharing them among multiple maps, but I'm not sure that that is the case anymore, so might be fine where they are?

File: src/components/templates/PartyMap/components/Map/Map.tsx
Comment (0xdevalias) [inline src/components/templates/PartyMap/components/Map/Map.tsx:87]: Can we refactor this to use `undefined` instead of `null`?

File: src/components/templates/PartyMap/components/Map/Map.tsx
Comment (0xdevalias) [inline src/components/templates/PartyMap/components/Map/Map.tsx:189]: > this probably doesn't even need to be a hook.. it's more of a component if anything. We can clean this up later

File: src/components/templates/PartyMap/PartyMap.tsx
Comment (0xdevalias) [inline src/components/templates/PartyMap/PartyMap.tsx:36]: > can we just make this a boolean that is set from isTruthy(selectedRoom)?

Apparently the nuance of how this is used is used to neatly show/hide the `RoomModal` without it 'losing the data' as it's transitioning to hidden. Not sure if that is a big deal though?

File: src/components/templates/PartyMap/PartyMap.tsx
Comment (0xdevalias) [inline src/components/templates/PartyMap/PartyMap.tsx:49]: > extract this into a reusable hook/similar

File: src/components/templates/PartyMap/PartyMap.tsx
Comment (0xdevalias) [inline src/components/templates/PartyMap/PartyMap.tsx:56]: > we could process this once to make it look uppable directly? What does the data key of venues look like?

File: src/types/venues.ts
Comment (0xdevalias) [reply src/types/venues.ts]: fixed in https://github.com/sparkletown/sparkle/pull/1080/commits/58906691ed0baa57fa53bf359d66839f100e6c5d

File: src/components/molecules/NavSearchBar/NavSearchBar.tsx
Comment (0xdevalias) [reply src/components/molecules/NavSearchBar/NavSearchBar.tsx]: fixed in https://github.com/sparkletown/sparkle/pull/1080/commits/12d266eba2172cc43eefd92518effc070d2cf1f1

File: src/api/profile.ts
Comment (0xdevalias) [reply src/api/profile.ts:20]: I think this would require some data backfilling/otherwise handling things differently.. so feels OOS for this PR

File: src/components/templates/PartyMap/components/Map/Map.tsx
Comment (0xdevalias) [reply src/components/templates/PartyMap/components/Map/Map.tsx:87]: I think for the same reason as https://github.com/sparkletown/sparkle/pull/1080#discussion_r559946760 this is probably OOS for this PR

File: src/components/templates/PartyMap/PartyMap.tsx
Comment (denisdimitrov) [reply src/components/templates/PartyMap/PartyMap.tsx:36]: Can you use the `selectedRoom` instead of the state above to show/hide the modal?

**Review by denisdimitrov — DISMISSED** [review]
LGTM!

File: src/components/templates/PartyMap/PartyMap.tsx
Comment (0xdevalias) [reply src/components/templates/PartyMap/PartyMap.tsx:36]: I thought about that as well, but then I realised that some of the little UI bugs i'm seeing at the moment (eg. the modal closing instantly rather than 'smoothly') is because I joined these concepts together, and when I unset `selectedRoom`, the modal doesn't render anything.

I had some thoughts about how I could fix this, but need to get the 🧠  back into this code to check if they'll work, or if I just need to revert this to being two separate things entirely.

Comment (0xdevalias) [reply src/components/templates/PartyMap/PartyMap.tsx]: fixed in https://github.com/sparkletown/sparkle/pull/1080/commits/2c8cd7105dd5d01c7b58c1f7f9751c3fb7022ba5

**Comment by 0xdevalias** [comment]
From https://github.com/sparkletown/sparkle/pull/1080#issuecomment-762615593:

> * [ ]  the border radius for the room image is missing
>   
>   * @denisdimitrov What are you referring to by this? In looking at the generated styles in `staging` there didn't seem to be a border radius on the room image. Can you please point out more specifically where you mean?

@denisdimitrov are you able to elaborate on the border radius you were referring to? I'm not sure what you mean.

**Comment by 0xdevalias** [comment]
@denisdimitrov can I please get you to re-review this. There were a bunch of conflicts I had to merge in again when I updated this against `staging`, and want to get another set of 👀  to ensure I didn't accidentally lose/regress anything with it.

File: src/components/molecules/NavSearchBar/NavSearchBar.tsx
Comment (mike-lvov) [inline src/components/molecules/NavSearchBar/NavSearchBar.tsx:62]: `roomsResults: Room[] =`

File: src/components/templates/AvatarGrid/AvatarGrid.tsx
Comment (mike-lvov) [inline src/components/templates/AvatarGrid/AvatarGrid.tsx:25]: [OOS] I think this template should be deleted

File: src/components/templates/PartyMap/components/Map/hooks/useMapGrid.tsx
Comment (mike-lvov) [inline src/components/templates/PartyMap/components/Map/hooks/useMapGrid.tsx:43]: What is the reason for changing it from `!!seatedPartygoer` to `eatedPartygoer !== undefined`?

File: src/components/templates/PartyMap/components/RoomAttendance.tsx
Comment (mike-lvov) [inline src/components/templates/PartyMap/components/RoomAttendance.tsx:28]: I think the logic was changed here. Was it tested?

File: src/components/templates/PartyMap/components/RoomModal/RoomModal.tsx
Comment (mike-lvov) [inline src/components/templates/PartyMap/components/RoomModal/RoomModal.tsx:24]: Not sure if we want to leave it as `..`. If so, should we also stick to `.`, instead of `./`?

File: src/components/templates/Playa/PlayaIcon.tsx
Comment (mike-lvov) [inline src/components/templates/Playa/PlayaIcon.tsx:4]: I thought Playa is completely removed from our codebase?

File: src/components/templates/Playa/PlayaVenue.tsx
Comment (mike-lvov) [inline src/components/templates/Playa/PlayaVenue.tsx:4]: I thought Playa is completely removed from our codebase?

File: src/components/templates/Playa/Router.tsx
Comment (mike-lvov) [inline src/components/templates/Playa/Router.tsx:8]: I thought Playa is completely removed from our codebase?

File: src/components/templates/Playa/VenuePreview.tsx
Comment (mike-lvov) [inline src/components/templates/Playa/VenuePreview.tsx:8]: I thought Playa is completely removed from our codebase?

File: src/pages/Admin/AdminVenuePreview.tsx
Comment (mike-lvov) [inline src/pages/Admin/AdminVenuePreview.tsx:70]: This type should move to `props` instead of `as` assignment

**Review by mike-lvov — CHANGES_REQUESTED** [review]
A handful of style-related comments. Some changes would be really nice to have in this PR

File: src/components/templates/Playa/VenuePreview.tsx
Comment (0xdevalias) [reply src/components/templates/Playa/VenuePreview.tsx:8]: It will be in the PR that removes it, which is separate to this PR, and will be merged after this PR. Both have been hanging around for ages, and this PR is the blocker (in my mental ordering/desire to complete things) of me working on the ‘delete all the old things’ PR.

File: src/components/templates/Playa/Router.tsx
Comment (0xdevalias) [reply src/components/templates/Playa/Router.tsx:8]: See https://github.com/sparkletown/sparkle/pull/1080#discussion_r564114898

File: src/components/templates/Playa/PlayaVenue.tsx
Comment (0xdevalias) [reply src/components/templates/Playa/PlayaVenue.tsx:4]: See https://github.com/sparkletown/sparkle/pull/1080#discussion_r564114898

File: src/components/templates/Playa/PlayaIcon.tsx
Comment (0xdevalias) [reply src/components/templates/Playa/PlayaIcon.tsx:4]: See https://github.com/sparkletown/sparkle/pull/1080#discussion_r564114898

File: src/components/templates/PartyMap/components/Map/hooks/useMapGrid.tsx
Comment (0xdevalias) [reply src/components/templates/PartyMap/components/Map/hooks/useMapGrid.tsx:43]: A general preference towards being more explicit in our intent with things, as that reduces the cognitive effort required to reason about it later. Potentially also a scope narrowing thing.

I don’t think I feel particularly strongly about this particular case either way, and I believe this particular field may also be able to be ‘null’ as well (as much as I dislike that it can be), so this change may not be accurate anyway.

File: src/components/templates/AvatarGrid/AvatarGrid.tsx
Comment (0xdevalias) [reply src/components/templates/AvatarGrid/AvatarGrid.tsx:25]: See https://github.com/sparkletown/sparkle/pull/1080#discussion_r564114898

File: src/components/templates/PartyMap/components/RoomAttendance.tsx
Comment (0xdevalias) [reply src/components/templates/PartyMap/components/RoomAttendance.tsx:28]: (as a general side note, it’s far easier to see what you are referring to if you select all of the relevant lines. In this case, all I can see is the deleted lines, and so need to go and open the file/find this section/etc to be able to comment further)

Comment (0xdevalias) [reply src/components/templates/PartyMap/components/RoomAttendance.tsx:28]: Previously it would just render nothing at all when there are no users in the room. Now it will still render the ‘container div’, but just won’t have anything in it.

I believe I did visually test this and it seemed to look fine (though for reasons that I believe are outside of the scope of this PR, I don’t think I was ever seeing any users in this room list, which I assume has/will be fixed with your users related PR, and/or will make more sense to track down/solve as a new PR once this + your users PR are merged)

File: src/components/templates/PartyMap/components/RoomModal/RoomModal.tsx
Comment (0xdevalias) [reply src/components/templates/PartyMap/components/RoomModal/RoomModal.tsx:24]: I don't actually have a strong opinion one way or the other about this. I wonder if there are best practices/linting rules that prefer one over the other?

File: src/components/molecules/NavSearchBar/NavSearchBar.tsx
Comment (0xdevalias) [reply src/components/molecules/NavSearchBar/NavSearchBar.tsx:62]: I believe the intent of your comment here is to then be able to remove the `roomsResults as Room[]` below. In which case, I agree.

Comment (0xdevalias) [reply src/components/molecules/NavSearchBar/NavSearchBar.tsx:62]: Looking at this, it is complicated because `venue?.rooms?` is `Room[] | AnyRoom[]`, which is probably why i did the `as` in the first place.

I'll add a `@debt` comment, but probably won't try and change it in this PR.

I believe once we delete all of the old stuff in https://github.com/sparkletown/sparkle/pull/833 we should then be able to clean this up properly.

File: src/components/templates/PartyMap/components/Map/hooks/useMapGrid.tsx
Comment (0xdevalias) [reply src/components/templates/PartyMap/components/Map/hooks/useMapGrid.tsx:43]: Looking at the typing here, that doesn't seem to be any need for `null` to exist? So will probably leave this as is unless there's a good use case why we shouldn't somewhere (eg. the typing lying to us)

Screen Shot 2021-01-27 at 6 58 06 pm

File: src/pages/Admin/AdminVenuePreview.tsx
Comment (0xdevalias) [reply src/pages/Admin/AdminVenuePreview.tsx:70]: Screen Shot 2021-01-27 at 7 01 22 pm

Looks like it was done this way to work around TypeScript. There would be ways we could 'fix this properly', but since this is existing code that i've just updated the types on, I think I will call it OOS for this PR.

I expect that once we can delete all of the old code in https://github.com/sparkletown/sparkle/pull/833, it will be simpler to see how we can properly fix/unify/simplify our types.

Comment (0xdevalias) [reply src/pages/Admin/AdminVenuePreview.tsx:70]: It's also possible that we might be able to refactor `PartyMapContainer` to be able to take `WithId<Venue>` directly.. which would solve this here. Will check if that breaks anything then comment back here.

Comment (0xdevalias) [reply src/pages/Admin/AdminVenuePreview.tsx:70]: Looks like the changes are going to be annoying to make, and it's not a big enough issue to be worth blocking this PR over (as it's more of a typing issue than a 'real' issue, as we already check that it's one of the correct template types in the `case` statement before this code runs.

So going to call this OOS for this PR.

**Review by mike-lvov — APPROVED** [review]
LGTM

**Pull request merged**
